Import monitor tool from DragonflyBSD

ricardobranco777 opened the pull request:
- Import [monitor](https://leaf.dragonflybsd.org/cgi/web-man?command=monitor&section=ANY) tool from DragonflyBSD.
- Add support for NOTE_OPEN, NOTE_CLOSE, NOTE_CLOSE_WRITE & NOTE_READ supported by FreeBSD & NetBSD 10.
- Add support for NOTE_TRUNCATE supported by OpenBSD.

Also trying with FreeBSD: https://github.com/freebsd/freebsd-src/pull/1087

ricardobranco777 committed:
Import monitor tool from DragonflyBSD

thorpej commented on usr.bin/monitor/monitor.c:
```diff
@@ -0,0 +1,248 @@
+	EV_SET(&kev, elm->fd, EVFILT_VNODE, EV_ADD | EV_ENABLE | EV_CLEAR,
+		NOTE_DELETE | NOTE_WRITE | NOTE_EXTEND | NOTE_ATTRIB |
+		NOTE_LINK | NOTE_RENAME | NOTE_REVOKE |
+#if defined(__FreeBSD__) || defined(__NetBSD__)
```
It would be better to `#ifdef` on the flag being tested rather than on the operating system.

ricardobranco777 replied:
> It would be better to `#ifdef` on the flag being tested rather than on the operating system.

In that case I think an `#ifdef NOTE_OPEN` should be enough. WDYT?

ricardobranco777 replied:
> It would be better to `#ifdef` on the flag being tested rather than on the operating system.

Removed the ifdef's as they pollute the code.

riastradh commented on usr.bin/monitor/monitor.c:
```diff
@@ -0,0 +1,235 @@
+static monitor_elm_t *Elms;
+
+int
+main(int ac, char **av)
```
argc, argv

riastradh commented on usr.bin/monitor/monitor.c:
```diff
@@ -0,0 +1,235 @@
+typedef struct monitor_elm {
+	const char *path;
+	int	fd;
+} *monitor_elm_t;
```
no typedef

https://nxr.netbsd.org/xref/src/share/misc/style?r=1.77#78

riastradh commented on usr.bin/monitor/monitor.c:
```diff
@@ -0,0 +1,235 @@
+} *monitor_elm_t;
+
+static void usage(int exit_code);
+static void monitor_add(const char *path);
```
omit parameter names from prototype

https://nxr.netbsd.org/xref/src/share/misc/style?r=1.77#205

riastradh commented on usr.bin/monitor/monitor.c:
```diff
@@ -0,0 +1,235 @@
+#include <sys/event.h>
+#include <sys/time.h>
+#include <sys/stat.h>
+#include <stdio.h>
```
blank line between include groups

riastradh commented on usr.bin/monitor/monitor.c:
```diff
@@ -0,0 +1,235 @@
+/*
```
add `/*	$NetBSD$	*/` on first line

riastradh commented on usr.bin/monitor/monitor.1:
```diff
@@ -0,0 +1,69 @@
+.\" Copyright (c) 2008 The DragonFly Project.  All rights reserved.
```
add `.\"	$NetBSD$` on first line

riastradh commented on usr.bin/monitor/monitor.c:
```diff
@@ -0,0 +1,235 @@
+
+	if ((KQueueFd = kqueue()) < 0) {
+		perror("kqueue");
+		exit(1);
```
err(EXIT_FAILURE, "kqueue")

riastradh commented on usr.bin/monitor/monitor.c:
```diff
@@ -0,0 +1,235 @@
+	int n;
+
+	elm = malloc(sizeof(*elm));
+	bzero(elm, sizeof(*elm));
```
calloc(1, sizeof(*elm)), and check for failure

(use explicit_memset, not bzero, where necessary)

riastradh commented on usr.bin/monitor/monitor.c:
```diff
@@ -0,0 +1,235 @@
+		exit(1);
+	}
+	NumFiles = MaxFiles = 16;
+	Elms = calloc(MaxFiles, sizeof(monitor_elm_t));
```
1. Why not just set NumFiles = MaxFiles = argc at this point, so there's no need to realloc anything?
2. Why an array of pointers and not an array of structs, so there's no need to malloc any elements?
3. Check for calloc failure.

riastradh commented on usr.bin/monitor/monitor.c:
```diff
@@ -0,0 +1,235 @@
+	bzero(elm, sizeof(*elm));
+	elm->path = path;
+	elm->fd = open(path, O_RDONLY);
+	if (elm->fd < 0) {
```
elm->fd == -1

riastradh commented on usr.bin/monitor/monitor.c:
```diff
@@ -0,0 +1,235 @@
+	elm->path = path;
+	elm->fd = open(path, O_RDONLY);
+	if (elm->fd < 0) {
+		printf("%s\tnot found\n", path);
```
Consider using warn(3), and/or print errno, and/or print strerror(errno)?

Showing the error is likely to be useful to the user.  Not sure whether it's best to go to stderr, as warn(3) does, or to produce a reliable parseable format that is easy to process with awk here, but one or the other would be good.

riastradh commented on usr.bin/monitor/monitor.c:
```diff
@@ -0,0 +1,235 @@
+
+	if (elm->fd >= NumFiles) {
+		MaxFiles = (elm->fd + 16) * 3 / 2;
+		Elms = realloc(Elms, MaxFiles * sizeof(elm));
```
reallocarr, not realloc

riastradh commented on usr.bin/monitor/monitor.c:
```diff
@@ -0,0 +1,235 @@
+	for (i = 0; i < n; ++i) {
+		kev = &kev_array[i];
+		elm = Elms[kev->ident];
+		printf("%-23s", elm->path);
```
I think it would be much more useful if this passed elm->path through vis(3) -- that way, it can be reliably passed through awk and parsed downstream.

It would make the output format different from dragonflybsd's tool, but I think making it composable with awk and other programs is worth that, especially if it gets renamed as others are suggesting.

riastradh commented on usr.bin/monitor/monitor.c:
```diff
@@ -0,0 +1,235 @@
+			printf(" ");
+			--bno;
+			kev->fflags &= ~(1 << bno);
+			switch (1 << bno) {
```
Use `__BIT(bno)` so this doesn't trigger undefined behaviour when bno=31 (or if fflags is ever expanded to uint64_t, which would also require using ffs64).

riastradh commented on usr.bin/monitor/monitor.c:
```diff
@@ -0,0 +1,235 @@
+				printf("read");
+				break;
+			default:
+				printf("%08x", 1 << bno);
```
`printf("%08x", (uint32_t)__BIT(bno))`

riastradh commented on usr.bin/monitor/monitor.c:
```diff
@@ -0,0 +1,235 @@
+usage(int exit_code)
+{
+	fprintf(stderr,
+		"monitor [-vx] files...\n"
```
four-space continuation lines

https://nxr.netbsd.org/xref/src/share/misc/style?r=1.77#367

riastradh replied:
missing -q in synopsis

riastradh commented on usr.bin/monitor/monitor.c:
```diff
@@ -0,0 +1,235 @@
+	if (elm->fd >= NumFiles) {
+		MaxFiles = (elm->fd + 16) * 3 / 2;
+		Elms = realloc(Elms, MaxFiles * sizeof(elm));
+		bzero(&Elms[NumFiles], (MaxFiles - NumFiles) * sizeof(elm));
```
explicit_memset

riastradh commented on usr.bin/monitor/monitor.c:
```diff
@@ -0,0 +1,235 @@
+		kev = &kev_array[i];
+		elm = Elms[kev->ident];
+		printf("%-23s", elm->path);
+		if (VerboseOpt && fstat(kev->ident, &st) == 0 &&
+		    S_ISREG(st.st_mode)) {
+			printf(" %10jd", (intmax_t)st.st_size);
+		}
```
Can this be presented differently so the format is more easily parseable?

ricardobranco777 replied:
> Can this be presented differently so the format is more easily parseable?

Any suggestion.  The size is always shown in a predictable field for awk.

riastradh reviewed (changes requested):
Cool, thanks, this seems like a good idea!  I noted some minor style nits, and I have one substantive suggestion: pass the paths through vis(3) and generally make the output format reliably parseable, so that it can be safely fed through awk or otherwise reliably ingested by other programs.  What do you think?

ricardobranco777 committed:
monitor: style fixes

ricardobranco777 committed:
monitor: Avoid reallocations

ricardobranco777 committed:
monitor: Use strvis for paths

ricardobranco777 commented:
> Cool, thanks, this seems like a good idea! I noted some minor style nits, and I have one substantive suggestion: pass the paths through vis(3) and generally make the output format reliably parseable, so that it can be safely fed through awk or otherwise reliably ingested by other programs. What do you think?

I think I addressed all your suggestions.  The only bug though is that a file should no longer be monitored after a delete event, since it no longer catches events after that.

I no longer think this tool belongs at all to base but examples instead.
